cli/asc.js

```diff
@@ -625,7 +625,7 @@ exports.main = function main(argv, options, callback) {
     if (numErrors) {
       const err = Error(numErrors + " parse error(s)");
       err.stack = err.message; // omit stack
-      return callback(err);
+      return callback(err, program.diagnostics);
```

Comment:
When compiling the compiler to Wasm, `program.diagnostics` is just a pointer into memory, making this non-portable. Perhaps, an alternative could be to add a callback to `APIOptions` that receives all the concrete data as arguments, not just a pointer, with the compiler only translating the diagnostics from memory to JS objects if the callback is present?

Reply:
Makes sense :+1: It could be named `reportDiagnostic` to follow [TypeScript naming](https://github.com/microsoft/TypeScript/blob/716b167f2b19e1642bd65aba33400afe8c181889/src/compiler/watchPublic.ts#L196). I will update it having WASM portability in mind :)

Reply:
How can I test it with the WASM version? I was able to build it with `asbuild` script from package.json - I have the `out` directory with binaries. I don't know what to do next :)

Reply:
The Wasm variant in master is only partially functional currently. But the [stw](https://github.com/AssemblyScript/assemblyscript/tree/stw) branch, which is scheduled to land soon-ish, has a fully functional bootstrap. You could base this PR on top of the stw branch, and rebase once it lands. Instructions to run the bootstrapped compiler in the stw branch are:

```
npm run bootstrap:untouched
node bin/asc --wasm out/assemblyscript.untouched-bootstrap.wasm ...
```

Note that some internals changed there, with the changes documented in [this PR](https://github.com/AssemblyScript/website/pull/39) scheduled to land along the branch.

Reply:
I've updated the PR and tested it with regular and WASM version :) I'm not sure if we want to have all this code, that copies diagnostic from WASM, in `asc.js`, or if we want to move some of the logic to the assemblyscript binary.

Reply:
Hopes are that we can get rid of the translation layer eventually, once Wasm can work with objects and pass them between the two worlds. Until then, if we can make this simpler somehow, that'd also be good to do, but typically quickly goes into the direction of generating JS glue code. Has its pros and cons.